updater/internal/install/service_windows.go

```diff
@@ -0,0 +1,276 @@
+	// productName is the name of the installed product
+	productName string
+}
+
+// Start the service
+func (w windowsService) Start() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("failed to connect to service manager: %w", err)
+	}
+	defer m.Disconnect()
+
+	s, err := m.OpenService(w.serviceName)
+	if err != nil {
+		return fmt.Errorf("failed to open service: %w", err)
+	}
+	defer s.Close()
+
+	if err := s.Start(); err != nil {
+		return fmt.Errorf("failed to start service: %w", err)
+	}
+
+	return nil
+}
+
+// Stop the service
+func (w windowsService) Stop() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("failed to connect to service manager: %w", err)
+	}
+	defer m.Disconnect()
+
+	s, err := m.OpenService(w.serviceName)
+	if err != nil {
+		return fmt.Errorf("failed to open service: %w", err)
+	}
+	defer s.Close()
+
+	if _, err := s.Control(svc.Stop); err != nil {
+		return fmt.Errorf("failed to start service: %w", err)
+	}
+
+	return nil
+}
+
+// Installs the service
+func (w windowsService) Install() error {
+	wsc, err := readWindowsServiceConfig(w.newServiceFilePath)
+	if err != nil {
+		return fmt.Errorf("failed to read service config: %w", err)
+	}
+
+	iDir, err := installDirFromRegistry(w.productName)
+	if err != nil {
+		return fmt.Errorf("failed to get install dir: %w", err)
+	}
+
+	expandArguments(wsc, w.productName, iDir)
+
+	splitArgs, err := shellquote.Split(wsc.Service.Arguments)
+	if err != nil {
+		return fmt.Errorf("failed to parse arguments in service config: %w", err)
+	}
+
+	startType, delayed, err := startType(wsc.Service.Start)
+	if err != nil {
+		return fmt.Errorf("failed to parse start type in service config: %w", err)
+	}
+
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("failed to connect to service manager: %w", err)
+	}
+	defer m.Disconnect()
+
+	s, err := m.CreateService(w.serviceName,
+		filepath.Join(iDir, wsc.Path),
+		mgr.Config{
+			Description:      wsc.Service.Description,
+			DisplayName:      wsc.Service.DisplayName,
+			StartType:        startType,
+			DelayedAutoStart: delayed,
+		},
+		splitArgs...,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to create service: %w", err)
+	}
+	defer s.Close()
+
+	return nil
+}
+
+// Uninstalls the service
+func (w windowsService) Uninstall() error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return fmt.Errorf("failed to connect to service manager: %w", err)
+	}
+	defer m.Disconnect()
+
+	s, err := m.OpenService(w.serviceName)
+	if err != nil {
+		return fmt.Errorf("failed to open service: %w", err)
+	}
+
+	if err = s.Delete(); err != nil {
+		sCloseErr := s.Close()
+		if sCloseErr != nil {
+			log.Default().Printf("Failed to close service: %s\n", err)
+		}
+		return fmt.Errorf("failed to delete service: %w", err)
+	}
+
+	if err := s.Close(); err != nil {
+		return fmt.Errorf("failed to close service: %w", err)
+	}
+
+	// Wait for the service to actually be deleted:
+	for {
```

Comment:
Is there a risk we get stuck in this loop? Guessing not because of the break condition if the service does not exist.

Reply:
This is a good question; I'm trying to consider what to do here. Really, this COULD take forever if some process just keeps a handle to this service open forever. So we should definitely guard against it.

I think we might want a timeout mechanism for this scenario, specifically. I think we might need to pass in a context here. Probably for all these service functions.

Reply:
I'm getting deja-vu now. I think we've run into this problem before, and I think we did not find a good solution to the problem.

If we mark the service for deletion, then timeout -- the service is still marked for deletion, and we can't undo that, and we'll end up in trouble.

But maybe we actually don't need to uninstall + install, maybe we can instead just "update" the service instead.

Reply:
I think just wait for it if we need to delete the service. It will eventually delete might as well wait for it to.